mantidproject/mantid
# Fix problem with MCInteractionVolume track statistics when running multiple threads

**Description of work.**

The debug logging of statistics in MCInteractionVolume wasn't working when the code was running from the MonteCarloAbsorption algorithm with multiple threads
Fixed this by creating a MCInteractionStatistics class and passing an object of this class into MCInteractionVolume instead of logging the stats on fields within MCInteractionVolume

As part of this I have reinstated the const qualifier on the m_scatterVol member of MCAbsorptionStrategy that was removed when the logging was first introduced a few months ago

**To test:**

Run this script and observe that the debug logging shows 100 scatter points generated for each detector. The script generates absorption corrections for 9 detectors on Pearl. Previously the total scatter point counts reported were less than 100 due to the threads interfering with each other:

```
from __future__ import print_function, division, unicode_literals

from mantid.simpleapi import (ConvertUnits, Load, LoadSampleShape, MonteCarloAbsorption, SetSample, CreateGroupingWorkspace, GroupDetectors)
import matplotlib.pyplot as plt

# Grab some data. This happens to be a run with a 10mm vanadium sphere

# select spectra for detector ids 1040, 2040, 3040, 4040, 5040, 6040, 7040, 8040, 9040
sample_env_mesh = Load(Filename='PEARL00073987.raw', SpectrumList='48,126,204,282,360,438,516,594,672')
# MonteCarloAbsorption requires an input ws with x units of wavelength - this determines
# which wavelengths to calculate the absorption correction for
sample_env_mesh = ConvertUnits(InputWorkspace=sample_env_mesh, Target='Wavelength')

SetSample(InputWorkspace=sample_env_mesh, Environment={'Name':'Pearl','Container':'Gasket_inner'},Material={'ChemicalFormula': 'Ni'})

sample_env_mesh_abscorr = MonteCarloAbsorption(InputWorkspace=sample_env_mesh, EventsPerPoint=100)
```
The debug logging should contain text like this for each detector:
```
Scatter point statistics
========================
Detector ID: 8040
Total scatter points generated: 100
Total scatter points used: 100
Sample: 4 (4.00%)
Environment part 0 (Gasket_inner): 2 (2.00%)
Environment part 1 (B-sdough): 0 (0.00%)
Environment part 2 (B-sdough): 0 (0.00%)
Environment part 3 (Anvil): 43 (43.00%)
Environment part 4 (Anvil): 50 (50.00%)
Environment part 5 (Gasket_outer): 0 (0.00%)
Environment part 6 (Gasket_inner): 1 (1.00%)
Environment part 7 (Fret): 0 (0.00%)
Environment part 8 (Fret): 0 (0.00%)
Scattering angle mean (degrees)=90.14
Scattering angle sd (degrees)=0.51
```

Fixes #28794.

*This does not require release notes* because it fixes a bug introduced since the 5.0 release in #28743

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
